### PR TITLE
Initial version of Token2

### PR DESCRIPTION
**Pull request by Kanapriya**

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        }
+    }
+
+    /**
+     * Get the Token2 user id
+     * @param context The authentication context
+     * @return Token2 userId value
+     *
+     */
+
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Remove this empty line.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        }
+        if (StringUtils.isEmpty(url)) {
+            return "";
+        } else {
+            try {
+                URL obj = new URL(url + Token2Constants.API + Token2Constants.EQUAL + apiKey + Token2Constants.AMPERSAND +
+                        Token2Constants.TOKEN + Token2Constants.EQUAL + token + Token2Constants.AMPERSAND +
+                        Token2Constants.USER_ID + Token2Constants.EQUAL + userId + Token2Constants.AMPERSAND +
+                        Token2Constants.FORMAT + Token2Constants.EQUAL + format);
+                HttpURLConnection urlConnection = (HttpURLConnection) obj.openConnection();
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Don't we need to close this connection?

File: component/authenticator/src/main/resources/token2.jsp [hunk 3]
```diff
@@ -0,0 +1,157 @@
+<%--
```
Comment (keerthu) [inline component/authenticator/src/main/resources/token2.jsp]: You don't need this class. Remove it.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        }
+    }
+
+    /**
+     * Get the Token2 user id
+     *
+     * @param context The authentication context
+     * @return Token2 userId value
+     */
+
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Remove this empty line.

File: component/authenticator/pom.xml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Private-Package>org.wso2.carbon.identity.authenticator.token2.internal</Private-Package>
+                        <Import-Package>
+                            org.wso2.carbon.identity.mgt.*; version="${carbon.identity.package.import.version.range}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: The versions are get from parent pom properties.

File: component/authenticator/pom.xml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                            org.apache.commons.logging.*; version="1.0.4",
+                            org.osgi.framework,
+                            org.wso2.carbon.identity.application.authentication.framework.*,
+                            javax.servlet,
+                            javax.servlet.http,
+                            org.apache.oltu.oauth2.*; version="${oltu.package.import.version.range}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: The versions are get from parent pom properties.

File: component/authentication-endpoint/src/main/webapp/localstyles.css [hunk 7]
```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2014, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (Kanapriya) [inline component/authentication-endpoint/src/main/webapp/localstyles.css]: Fixed https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authentication-endpoint/src/main/webapp/localstyles.css [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    background: #45484d; /* Old browsers */
+    background: -moz-linear-gradient(top, #45484d 0%, #000000 100%); /* FF3.6+ */
+    background: -webkit-gradient(linear, left top, left bottom, color-stop(0%, #45484d), color-stop(100%, #000000)); /* Chrome,Safari4+ */
+    background: -webkit-linear-gradient(top, #45484d 0%, #000000 100%); /* Chrome10+,Safari5.1+ */
+    background: -o-linear-gradient(top, #45484d 0%, #000000 100%); /* Opera 11.10+ */
+    background: -ms-linear-gradient(top, #45484d 0%, #000000 100%); /* IE10+ */
+    background: linear-gradient(to bottom, #45484d 0%, #000000 100%); /* W3C */
+    filter: progid:DXImageTransform.Microsoft.gradient(startColorstr='#45484d', endColorstr='#000000', GradientType=0); /* IE6-9 */
+    padding: 5px 10px;
+
```
Comment (Kanapriya) [inline component/authentication-endpoint/src/main/webapp/localstyles.css]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+import java.net.MalformedURLException;
+import java.net.ProtocolException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Authenticator of Token2
+ */
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    /**
+     * Get the configuration properties of UI
+     */
+    @Override
+    public List<Property> getConfigurationProperties() {
+        List<Property> configProperties = new ArrayList<>();
+
+        Property apiKey = new Property();
+        apiKey.setName(Token2Constants.APIKEY);
+        apiKey.setDisplayName("Api Key");
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: fixed in https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+        apiKey.setRequired(true);
+        apiKey.setDescription("Enter Token2 API Key value");
+        apiKey.setDisplayOrder(1);
+        configProperties.add(apiKey);
+
+        Property callbackUrl = new Property();
+        callbackUrl.setDisplayName("Callback URL");
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+        callbackUrl.setName(IdentityApplicationConstants.OAuth2.CALLBACK_URL);
+        callbackUrl.setDescription("Enter value corresponding to callback url.");
+        callbackUrl.setDisplayOrder(2);
+        configProperties.add(callbackUrl);
+        return configProperties;
+    }
+}
+
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Constants.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    public static final String VALIDATION = "validation";
+    public static final String FORMAT = "format";
+    public static final String GET_METHOD = "GET";
+    public static final String USERID_CLAIM = "http://wso2.org/claims/identity/userid";
+    public static final String EQUAL = "=";
+    public static final String AMPERSAND = "&";
+    public static final String LOGIN_PAGE = "token2authenticationendpoint/token2.jsp";
+    public static final String RETRY_PARAMS = "&authFailure=true&authFailureMsg=login.fail.message";
+    public static final String TOKEN2_VALIDATE_ENDPOINT = "https://api.token2.com/validate?";
+
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Constants.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authenticator/pom.xml (hunk 5 above)
Comment (Kanapriya) [inline component/authenticator/pom.xml]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-token2/pull/2

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+    /**
+     * Get the Token2 user id
+     *
+     * @param context The authentication context
+     * @return Token2 userId value
+     */
+    private String getUserId(AuthenticationContext context) throws AuthenticationFailedException {
+        String userId = null;
+        String username = getUsername(context);
+        if (username != null) {
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Use StringUtils.isEmpty() for null check.

Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/Token2Authenticator.java]: Fixed in 23b4e6a4cb69cf0f295f61cccdee84db0f76c28e

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+
+    protected void deactivate(ComponentContext ctxt) {
+        if (log.isDebugEnabled()) {
+            log.debug("Token2 authenticator is deactivated");
+        }
+    }
+    protected void setRealmService(RealmService realmService) {
+        log.debug("Setting the Realm Service");
```
Comment (dinusha-dilrukshi) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java]: Check isDebugEanbled().

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        Token2AuthenticatorServiceComponent.realmService = realmService;
+    }
+
+    protected void unsetRealmService(RealmService realmService) {
+        log.debug("UnSetting the Realm Service");
```
Comment (dinusha-dilrukshi) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java]: Check isDebugEanbled().

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java (hunk 15 above)
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java]: Fixed in 24c7505c9768ead90869a8435d2bc5be0060b89b

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java (hunk 16 above)
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/token2/internal/Token2AuthenticatorServiceComponent.java]: Fixed in 24c7505c9768ead90869a8435d2bc5be0060b89b

**Pull request merged**